text/0000-non-ascii-idents.md

```diff
@@ -0,0 +1,171 @@
+
+To disallow any Unicode identifiers in a project (for example to ease collaboration or for security reasons) limiting the accepted identifiers to ASCII add this lint to the `lib.rs` or `main.rs` file of your project:
+
+```rust
+#![forbid(non_ascii_idents)]
+```
+
+Some Unicode character look confusingly similar to each other or even identical like the Latin **A** and the Cyrillic **А**. The compiler may warn you about names that are easy to confuse with keywords, names from the same crate and imported items. If needed (but not recommended) this warning can be silenced with a `#[allow(confusable_idents)]` annotation on the enclosing function or module.
+
+## Usage notes
+
+All code written in the Rust Language Organization (*rustc*, tools, std, common crates) will continue to only use ASCII identifiers and the English language.
+
+For open source crates it is suggested to write them in English and use ASCII-only. An exception can be made if the application domain (e.g. math) benefits from Unicode and the target audience (e.g. for a crate interfacing with Russian passports) is comfortable with the used language and characters. Additionally crates should consider to provide an ASCII-only API.
+
+Private projects can use any script and language the developer(s) desire. It is still a good idea (as with any language feature) not to overdo it.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Identifiers in Rust are based on the [Unicode® Standard Annex #31 Unicode Identifier and Pattern Syntax][UAX31].
+
+Note: The supported Unicode version should be stated in the documentation.
+
+The lexer defines identifiers as:
+
+> **<sup>Lexer:<sup>**  
+> IDENTIFIER_OR_KEYWORD:  
+> &nbsp;&nbsp; XID_Start&nbsp;XID_Continue<sup>\*</sup>  
+> &nbsp;&nbsp; | `_` XID_Continue<sup>+</sup>  
+>  
+> IDENTIFIER :  
+> IDENTIFIER_OR_KEYWORD <sub>*Except a [strict] or [reserved] keyword*</sub>
+
+`XID_Start` and `XID_Continue` are used as defined in the aforementioned standard. The definition of identifiers is forward compatible with each successive release of Unicode as only appropriate new characters are added to the classes but none are removed.
+
+Parsers for Rust syntax normalize identifiers to [NFC][UAX15]. Every API accepting raw identifiers (such as `proc_macro::Ident::new` normalizes them to NFC and APIs returning them as strings (like `proc_macro::Ident::to_string`) return the normalized form. This means two identifiers are equal if their NFC forms are equal.
+
+A `non_ascii_idents` lint is added to the compiler. This lint is `allow` by default. The lint checks if any identifier in the current context contains a codepoint with a value equal to or greater than 0x80 (outside ASCII range). Not only locally defined identifiers are checked but also those imported from other crates and modules into the current context. 
```

Comment:
_Bringing my position inline so that anyone can reply to this directly._

I believe that by default we should:

* Allow a usable subset of unicode characters in identifiers (UAX #31 or a variation of it sounds reasonable) by default.
* Below there's already a proposal for a confusable idents lint, it should be warn by default.
* Instead of having a binary ASCII/all flag, we should treat the full unicode space as per script flags, allowing projects to specify (perhaps even _per scope_) a subset of allowable scripts (`latin`, `math`, `greek`, `hira`/`hiragana`, `ascii`/`reduced`/`limited`/`lets-party-like-its-1960`). If a list of allowable scripts are defined in a crate, encountering chars outside of them is a hard error. On this error, confusables that are outside of the currently allowed scripts but that have a similar representation in the current script (`𝜆` in `math` -> `λ` in `greek`), it should be pointed out with suggestions to either use the allowed script's char or to add the new script to the allowed list.

Reply:
Can you please state the reasoning behind point three? If the supporting arguments were already discussed also include the opinions of other people on the matter.

Reply:
Being liberal on the default is a contentious point with people bringing multiple positions. Myself I fall under the _lets be relatively liberal_ camp, with _maybe_ a concession to a warn by default lint when falling outside of ascii suggesting specifying allowed scripts. By doing it this way we minimize friction from non-ascii using users, while not heavily burdening ascii-only users (this is debated).

I believe we should allow people to specify specific allowed scripts is a good solution to the problem of cross script confusables sneaking in, without having people forced to decide that it is not worth it to use math symbols for formulas because they're are worried about cyrillic confusables being sneaked into the codebase. This would also allow me to use emojis for identifiers if I so choose, as that script would be disabled by default, but I could add it to the allowed list.

Allowing per scope customization would be interesting to allow people a lot of linting control when deciding that they want to allow a given script internally, but not on the public api. I would not count function argument names as part of the public api for this consideration, as we don't need to write them down, just be able to read them.

Reply:
Would you say yourself that this is a fair and complete review of the previous arguments?

Reply:
I believe it to be so, yes. I haven't seen my last point around per scope settings raised before, though.

Reply:
> How does allowing more codepoints than UAX#31 Default Identifiers the evolution and readability of Rust? (See custom operators and smart quotes)

I feel that giving partial support for Unicode idents would be a disservice if we don't allow a future proof escape hatch.

> For which people is it important to allow additional characters?

Other than novelty uses (emoji), I don't know enough about other languages that might need it (my mother tongue is limited to latin).

> Why is the proposed confusable lint insufficient?

Proactive warning and fine grained control for individual projects. A binary gate feels too blunt to mdgiven the scope.

> Why aren't ad hoc solutions like regular expressions sufficient?

I believe that the compiler and an editor should be enough to develop in Rust, and it should make an effort to provide help to newcomers for non obvious errors (and here I fall in the same camp of "avoid non ASCII in general" if possible, I just don't want to mandate it).

> Why are typos involving Unicode characters different from those only involving ASCII ones?

Tooling support. We already suggest based on levenshtein distance for ASCII typos, we should to the same for cobfusable chars.

Having said all of this, I'm enthusiastically in support of adding the proposed support and will do my best to make sure that the presented worries have a reasonable answer in the shape of compiler assisted support.

Reply:
> Having said all of this, I'm enthusiastically in support of adding the proposed support and will do my best to make sure that the presented worries have a reasonable answer in the shape of compiler assisted support.

Great :+1: 

> Other than novelty uses (emoji), I don't know enough about other languages that might need it (my mother tongue is limited to latin).

Maybe update your position above if it only for novelty?

> Proactive warning and fine grained control for individual projects. A binary gate feels too blunt to mdgiven the scope.

Care to give an example from your personal experience where such a feature in a programming language would have made your work easier? (Keep in mind that many languages have Unicode support for years now among them Python, Java and C++)

> Tooling support. We already suggest based on levenshtein distance for ASCII typos, we should to the same for cobfusable chars.

We already do (tested on nightly):

```
error: cannot find macro `príntln!` in this scope
 --> src/main.rs:4:5
  |
4 |     príntln!("hello world")
  |     ^^^^^^^ help: you could try the macro: `println`
```

Reply:
> Maybe update your position above if it only for novelty?

My position is born of my own ignorance, I defer to people with more experience with non-western-european languages. For me (again, other than the novelty use of emoji and math symbols), my use case can be completely covered by latin (ñ, ü, é) and my counter case would be identifying confusables (which as you point out it would be handled) and bad diagnostics when changed code (smart quotes) is pasted.

> Care to give an example from your personal experience where such a feature in a programming language would have made your work easier? (Keep in mind that many languages have Unicode support for years now among them Python, Java and C++)

Entire generation of Spanish speaking students have been "trained" to transliterate to ASCII, where `año` (year) gets changed to `anio` (meaningless) or `ano` (anus). I think that the focus for this document should be placed on the use of other scripts that limit developers more heavily than this.

[As pointed out](https://github.com/rust-lang/rfcs/pull/2457#issuecomment-396741621), having a single toggle would make it hard to only allow full unicode support at different levels, like only for literals, idents, in comments, or maybe on idents for internal interfaces, while disallowing them in other places (like public interfaces).